### PR TITLE
feat: add GreenletProfiler

### PR DESCRIPTION
**Pull request by westarest**

### CODE REVIEW

**Comment by westarest** [comment]
it is a draft, but it can work with threading model and greenlet model automatically now. it needs more things:
test cases

**Comment by wu-sheng** [comment]
Verifying through e2e profiling case would be a good place to start.

**Comment by wu-sheng** [comment]
Profiling e2e has been included in the main repo, https://github.com/apache/skywalking/blob/317d539371c588c4754029134ad76087ecb6d0cd/.github/workflows/skywalking.yaml#L435-L454 powered by our e2e-infra tool, https://skywalking.apache.org/docs/skywalking-infra-e2e/next/readme/

You could learn how we run tests using the latest agent with real OAP/SkyWalking backend, and verify the result through SkyWalking CLI through this tool.

**Comment by westarest** [comment]
> thank you, it is the docs what I need

**Comment by westarest** [comment]
there is something wrong when profiling start:
```
skywalking [Thread-127] [DEBUG] profile task [1666974131787_ZTJlLXNlcnZpY2UtcHJvdmlkZXI=.1] for endpoint [/artist] started
Traceback (most recent call last):
  File "src/gevent/_waiter.py", line 122, in gevent._gevent_c_waiter.Waiter.switch
  File "/skywalking-python/venv/lib/python3.6/site-packages/apache_skywalking-0.8.0-py3.6.egg/skywalking/profile/profile_context.py", line 329, in callback
    agent.add_profiling_snapshot(snapshot)
  File "/skywalking-python/venv/lib/python3.6/site-packages/apache_skywalking-0.8.0-py3.6.egg/skywalking/agent/__init__.py", line 288, in add_profiling_snapshot
    __snapshot_queue.put(snapshot)
  File "/usr/local/lib/python3.6/queue.py", line 126, in put
    with self.not_full:
  File "/usr/local/lib/python3.6/threading.py", line 240, in __enter__
    return self._lock.__enter__()
  File "src/gevent/_semaphore.py", line 278, in gevent._gevent_c_semaphore.Semaphore.__enter__
  File "src/gevent/_semaphore.py", line 279, in gevent._gevent_c_semaphore.Semaphore.__enter__
  File "src/gevent/_semaphore.py", line 180, in gevent._gevent_c_semaphore.Semaphore.acquire
  File "/skywalking-python/venv/lib/python3.6/site-packages/gevent/thread.py", line 121, in acquire
    acquired = BoundedSemaphore.acquire(self, blocking, timeout)
  File "src/gevent/_semaphore.py", line 180, in gevent._gevent_c_semaphore.Semaphore.acquire
  File "src/gevent/_semaphore.py", line 249, in gevent._gevent_c_semaphore.Semaphore.acquire
  File "src/gevent/_abstract_linkable.py", line 521, in gevent._gevent_c_abstract_linkable.AbstractLinkable._wait
  File "src/gevent/_abstract_linkable.py", line 487, in gevent._gevent_c_abstract_linkable.AbstractLinkable._wait_core
  File "src/gevent/_abstract_linkable.py", line 490, in gevent._gevent_c_abstract_linkable.AbstractLinkable._wait_core
  File "src/gevent/_abstract_linkable.py", line 442, in gevent._gevent_c_abstract_linkable.AbstractLinkable._AbstractLinkable__wait_to_be_notified
  File "src/gevent/_abstract_linkable.py", line 451, in gevent._gevent_c_abstract_linkable.AbstractLinkable._switch_to_hub
  File "src/gevent/_greenlet_primitives.py", line 61, in gevent._gevent_c_greenlet_primitives.SwitchOutGreenletWithLoop.switch
  File "src/gevent/_greenlet_primitives.py", line 64, in gevent._gevent_c_greenlet_primitives.SwitchOutGreenletWithLoop.switch
  File "src/gevent/_greenlet_primitives.py", line 67, in gevent._gevent_c_greenlet_primitives.SwitchOutGreenletWithLoop.switch_out
  File "src/gevent/_greenlet_primitives.py", line 68, in gevent._gevent_c_greenlet_primitives.SwitchOutGreenletWithLoop.switch_out
gevent.exceptions.BlockingSwitchOutError: Impossible to call blocking function in the event loop callback
2022-10-28T16:22:34Z <built-in method switch of gevent._gevent_cgreenlet.Greenlet object at 0x7ff213196a48> failed with BlockingSwitchOutError
```

File: .github/workflows/CI.yaml [hunk 1]
```diff
@@ -223,6 +223,10 @@ jobs:
             path: tests/e2e/case/http/e2e.yaml
           - name: Kafka
             path: tests/e2e/case/kafka/e2e.yaml
+          - name: proflinig_threading
```
Comment (Superskyyy) [inline .github/workflows/CI.yaml:226]:
```suggestion
          - name: profiling_threading
```

**Review by Superskyyy — CHANGES_REQUESTED** [review]

**Comment by Superskyyy** [comment]
Generally looks good to me. @kezhenxu94 can you take a look? I'm not very familiar with the profiling feature.

**Comment by Superskyyy** [comment]
The sonatype bug can be ignored, since it's testing code. LGTM.

**Review by Superskyyy — DISMISSED** [review]

**Comment by Superskyyy** [comment]
We need to update `changes.md` to reflect the new feature. And also we will need documentation. The original documentation for threading-profiling was posted to here https://skywalking.apache.org/blog/2021-09-12-skywalking-python-profiling/, now with the addition of Greenlet Profiler, it will be better to move it back to our repository and add the corresponding part for greenlet. @westarest, can you add the docs in the next PR?

File: skywalking/profile/profile_context.py [hunk 2]
```diff
@@ -34,6 +34,24 @@
...
 
 
+THREAD_MODEL = 'thread'
+try:
+    from gevent import monkey
+    import greenlet
+    from gevent.exceptions import BlockingSwitchOutError
+
+    if monkey.is_module_patched('threading'):
+        if greenlet.__version__ < '2.0.0':
```
Comment (kezhenxu94) [inline skywalking/profile/profile_context.py:44]: This looks error prone. Can you use the `packaging` module to parse and compare the version?

**Review by kezhenxu94 — DISMISSED** [review]

**Comment by westarest** [comment]
> We will need documentation. The original documentation for threading-profiling was posted to here https://skywalking.apache.org/blog/2021-09-12-skywalking-python-profiling/, now with the addition of Greenlet Profiler, it will be better to move it back to our repository and add the corresponding part for greenlet. @westarest, can you work on the docs in the next PR?

no problem. I can post a document about the difference between the Threading profiler  and the Greenlet profiler that I committed.

**Review by kezhenxu94 — APPROVED** [review]
Thanks!

**Review by wu-sheng — COMMENTED** [review]
Cool. @westarest Would you like to write a new blog for this feature, to introduce why it is important, and how you implement this, and use it?

I believe this is a very important and valuable blog.

**Pull request merged**